src/Microsoft.AspNet.Routing/VirtualPathData.cs

```diff
@@ -57,23 +56,23 @@ public VirtualPathData(IRouter router, string virtualPath)
 
             Router = router;
             VirtualPath = virtualPath;
-
-            _dataTokens = new RouteValueDictionary();
-            if (dataTokens != null)
-            {
-                foreach (var dataToken in dataTokens)
-                {
-                    _dataTokens.Add(dataToken.Key, dataToken.Value);
-                }
-            }
+            _dataTokens = dataTokens == null ? null : new RouteValueDictionary(dataTokens);
```

Comment:
The common case here is that there aren't any data tokens and that no one will look at them, so we're alloc-free when that happens. If you are using data tokens then pay-for-play